File: java/behavioral/observer/src/main/java/org/example/gof/behavioral/observer/StockTicker.java
package org.example.gof.behavioral.observer;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class StockTicker {
    private final List<Subscriber> subscribers = new ArrayList<>();
    private final Map<String, Double> lastPrices = new HashMap<>();

    public void subscribe(Subscriber subscriber) {
        subscribers.add(subscriber);
    }

    public void publish(String symbol, double price) {
        lastPrices.put(symbol, price);
        QuoteUpdate update = new QuoteUpdate(symbol, price);
        subscribers.forEach(subscriber -> subscriber.onNotify(update));
    }

    public Double lastPrice(String symbol) {
        return lastPrices.get(symbol);
    }
}

File: rust/behavioral/observer/examples/stock_ticker.rs
use observer::{HistoryLogger, StockTicker, Subscriber, QuoteUpdate};

struct AlertingSubscriber;

impl Subscriber for AlertingSubscriber {
    fn on_notify(&mut self, update: &QuoteUpdate) {
        println!("ALERT: {} is now {:.2}", update.symbol, update.price);
    }
}

fn main() {
    let mut ticker = StockTicker::default();
    let logger = HistoryLogger::new();
    ticker.subscribe(Box::new(AlertingSubscriber));
    ticker.subscribe(Box::new(logger));

    ticker.publish("MSFT", 421.3);
    ticker.publish("GOOGL", 176.5);
}
